### PR TITLE
Validate type in topology using current cache group name

### PR DESCRIPTION
**Pull request by ericholguin**
Closes: #6551 

Validates that cache group type is not being changed if it is used in a topology, using previous object name, instead of the requested name.

<hr/>

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

Try updating the type of a cache group which belongs to a topology, should fail
Try updating the type and name of a cache group which belongs to a topology, should fail

## If this is a bugfix, which Traffic Control versions contained the bug?

- 5.1.x
- 6.0.x
- 6.1.x

## PR submission checklist
- [x] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

**Comment by ericholguin** [comment]
This change will only fix the issue for Traffic Portal. The code performs a check against the requested body, so if in the request body there is a change to the name, the validation will perform a query against that new name and will not find that newly named cache group in any topology, hence allowing it to update and succeed. Changing the code to use the previous or current name, fixes this, as it will find the topology relationship.

However, in Traffic Ops, the type can be changed regardless of the name or any other field changes, if there is no id field in the request body. This is because the validation uses the cache group id in the query to get previous type and now this newly added query for previous name. The way validation is structured, is that it will only use the requested body, and in the case of this endpoint id is in the url path. If the user omits the id, the validation will be querying using an empty id value resulting in no real validation being made.

Not sure what the fix is, without changing the Validate functionality. Or unless id is required in the request body when updating a cache group.

**Review by zrhoffman — CHANGES_REQUESTED** [review]
> [](https://github.com/ericholguin)@[ericholguin](https://github.com/ericholguin)
Contributor Author
ericholguin commented [2 days ago](https://github.com/apache/trafficcontrol/pull/6552#issuecomment-1034361900) •

This change will only fix the issue for Traffic Portal. The code performs a check against the requested body, so if in the request body there is a change to the name, the validation will perform a query against that new name and will not find that newly named cache group in any topology, hence allowing it to update and succeed. Changing the code to use the previous or current name, fixes this, as it will find the topology relationship.

However, in Traffic Ops, the type can be changed regardless of the name or any other field changes, if there is no id field in the request body. This is because the validation uses the cache group id in the query to get previous type and now this newly added query for previous name. The way validation is structured, is that it will only use the requested body, and in the case of this endpoint id is in the url path. If the user omits the id, the validation will be querying using an empty id value resulting in no real validation being made.

Sounds like `id` should be required in PUT requests

File: traffic_ops/traffic_ops_golang/cachegroup/cachegroups.go [hunk 1]
```diff
@@ -188,14 +188,30 @@ func (cg *TOCacheGroup) ValidateTypeInTopology() error {
...
+	FROM cachegroup c
+	WHERE c.id = $1
+	`
+	var previousName string
+	err = cg.ReqInfo.Tx.QueryRow(previousNameQuery, *cg.ID).Scan(&previousName)
+	if err == sql.ErrNoRows {
+		return nil
+	}
+	if err != nil {
+		log.Errorf("%s: getting the previous name of cachegroup %s: %s", userErr.Error(), *cg.Name, err.Error())
```
Comment (zrhoffman) [inline traffic_ops/traffic_ops_golang/cachegroup/cachegroups.go:203]: This line panics  if `userErr` is nil. Changing `userErr.Error()` to `userErr` would fix that

**Review by zrhoffman — CHANGES_REQUESTED** [review]

File: traffic_ops/traffic_ops_golang/cachegroup/cachegroups.go (hunk 1 above)
Comment (ericholguin) [reply traffic_ops/traffic_ops_golang/cachegroup/cachegroups.go:203]: Doesn't this line prevent that?
https://github.com/apache/trafficcontrol/blob/c0b747c18799952a4db7566c5a0f5dd9a5989c95/traffic_ops/traffic_ops_golang/cachegroup/cachegroups.go#L164

Comment (zrhoffman) [reply traffic_ops/traffic_ops_golang/cachegroup/cachegroups.go:203]: oh, true

**Comment by zrhoffman** [comment]
> I'm not sure if adding that requirement would be a breaking change to the API versions or other implications since the Update Handler simply uses the url path to set the id key in the object

Good point, it would be.

**Pull request merged**